doc/config.md
# Config

The config.ini files are parsed by the Python `ConfigParser` and so follow the .ini file format.

## Model section
```ini
[model]
g_w = 8
k = 3
s = 2
hg_size = 128
hl_size = 128
g_size = 256
hidden_size = 256
glimpses = 6
num_classes = 10
loc_std = 0.1
```
- `g_w` : int
  + size of glimpse window in pixels, default is 8 x 8
- `k` : int
  + number of regions
- `s` : int
  + scaling factor
- `hg_size` : int
  + number of units in hidden layer that takes glimpse as input
- `hl_size` : int
  + number of units in hidden layer that takes glimpse location as input
- `g_size` : int
  + number of units in hidden layer that combines glimpse + location hidden layers
- `hidden_size` : int
  + number of units in core network
- `glimpses` : int
  + number of glimpses to take of each image
- `num_classes` : int
  + number of classes in data
- `loc_std` : float
  + standard deviation of distribution from which each glimpse location is drawn

## Data section
```ini
[data]
root_results_dir = /home/you/Documents/data/RAM_output
data_dir = /home/you/Documents/data/mnist/raw
module = ram.dataset.mnist
train_size = 0.9
val_size = 0.1
```
- `root_results_dir` : str
  + name of directory in which a new sub-directory will be created to hold results from run
- `data_dir` : str
  + name of directory where data is saved
- `module` : str
  + name of module that prepares and returns data sets, in absolute terms, i.e. 
   `package.module`, `package.subpackage.module`, etc. Python will use `importlib` to 
   import this module. This is a quick and dirty way to make it possible to use your 
   own data with the library. The module must implement two functions, `prep` and `get_split`.
   The function should accept arguments and return variables like so:
   ```Python
   paths_dict = dataset_module.prep(download_dir=config.data.data_dir,
                                    train_size=config.data.train_size,
                                    val_size=config.data.val_size,
                                    output_dir=config.data.data_dir)
   train_data, val_data = dataset_module.get_split(paths_dict, setname=['train', 'val'])
   ```
   For more detail and examples, see the `ram.__main__` function and the `ram.dataset.mnist` module.
- `val_size` : float
  + Amount of training data to use for a validation set
- `train_size` : float
  + Amount of entire training data set to use as a subset for training


## Train section
```ini
[train]
batch_size = 200
learning_rate = 1e-3
epochs = 200
optimizer = momentum
momentum = 0.9
checkpoint_prefix = ckpt
restore = False
save_log = Truesave_examples_every = 25
num_examples_to_save = 9
save_loss = True
save_train_inds = True
```
- `batch_size` : int
  + number of samples in each batch fed to the network
- `learning_rate` : float
  + learning rate used by optimizer during training
- `epochs` : int
  + number of training epochs, i.e. number of times network sees entire training set
- `optimizer` : str
  + one of {'adam', 'momentum', 'sgd'}; type of optimizer used
- `momentum`: float
  + if optimizer is 'momentum', the value for 'momentum'
- `checkpoint_prefix` : str
  + prefix for checkpoint files
- `restore` : bool
  + if True, restore from previous checkpoint. Not used by command-line interface but
  useful if you want to train your own models with the Trainer class.
- `save_examples_every` : int
  + if defined, save examples at an interval of this many epochs. Examples are
  glimpses seen by network, locations of gazes, fixations (locations converted from normalized to
  pixel co-ordinates), and images (from which glimpses were taken)
- 'num_examples_to_save' : int
  + number of examples to save, if `save_examples` is `True`
- 'save_loss' : bool
  if `True`, save records of loss from every epoch.
- 'save_train_inds' : bool
  if `True`, save the indices of the training samples used for each epoch.

## Test section
```ini
[test]
save_examples = True
num_examples_to_save = 15
```
- `save_examples` : bool
  + if True, save examples. Examples are
  glimpses seen by network, locations of gazes, fixations (locations converted from normalized to
  pixel co-ordinates), and images (from which glimpses were taken)
- 'num_examples_to_save' : int
  + number of examples to save, if `save_examples` is `True`

## Misc section
```ini
[misc]
save_log = True
```
- `save_log` : bool
  + if True, save logging to a text file.
- `random_seed` : int
  + number used to seed random number generator, to make results reproducible across runs


doc/CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]
### Added
- ability to use only a subset of MNIST training data and get a validation set from it
- ability to shuffle dataset on each epoch
- ability to train replicates (experiment repeated n times with same training data,
 only random initialization / shuffling different)
- ability to specify other modules to use to load other datasets
  + an example is provided to load the [kanji MNIST dataset](https://github.com/rois-codh/kmnist)
- logging of runs/experiments, with option to dump to a text file
- tests for MNIST module in datasets
- a CHANGELOG (this one)

### Changed
- change argparser to use positional arguments `command` and `config`
  + before all arguments were "optional" (although the program would crash without them)



## [0.1.0] - 2018-12-01
### Added
- original version, probably closest to the [one from Kevin Zakka](https://github.com/kevinzakka/recurrent-visual-attention)
- but in Tensorflow (Eager) instead of PyTorch


src/ram/config.py
import os
from configparser import ConfigParser

import attr

VALID_OPTIONS = {
    'model': [
        'g_w',
        'k',
        's',
        'hg_size',
        'hl_size',
        'g_size',
        'hidden_size',
        'glimpses',
        'num_classes',
        'loc_std'
    ],
    'train': [
        'batch_size',
        'learning_rate',
        'epochs',
        'optimizer',
        'momentum',
        'replicates',
        'checkpoint_prefix',
        'restore',
        'shuffle_each_epoch',
        'save_examples_every',
        'num_examples_to_save',
        'save_loss',
        'save_train_inds',

    ],
    'data': [
        'root_results_dir',
        'data_dir',
        'module',
        'train_size',
        'val_size',
        'results_dir_made_by_main'
    ],
    'test': [
        'save_examples',
        'num_examples_to_save',
    ],
    'misc': [
        'save_log',
        'random_seed'
    ]
}

VALID_SECTIONS = set(VALID_OPTIONS.keys())


# adapted from cPython/Lib/distutils/util.py
def strtobool(val):
    """Convert a string representation of truth to true (1) or false (0).
    True values are 'y', 'yes', 't', 'true', 'on', and '1'; false values
    are 'n', 'no', 'f', 'false', 'off', and '0'.  Raises ValueError if
    'val' is anything else.
    """
    val = val.lower()
    if val in ('y', 'yes', 't', 'true', 'on', '1'):
        return True
    elif val in ('n', 'no', 'f', 'false', 'off', '0'):
        return False
    else:
        raise ValueError("invalid truth value %r" % (val,))


@attr.s
class ModelConfig(object):
    """class that represents configuration for RAM model"""
    g_w = attr.ib(converter=int, default=8)
    k = attr.ib(converter=int, default=3)
    s = attr.ib(converter=int, default=2)
    hg_size = attr.ib(converter=int, default=128)
    hl_size = attr.ib(converter=int, default=128)
    g_size = attr.ib(converter=int, default=256)
    hidden_size = attr.ib(converter=int, default=256)
    glimpses = attr.ib(converter=int, default=6)
    num_classes = attr.ib(converter=int, default=10)
    loc_std = attr.ib(converter=float, default=0.1)


@attr.s
class DataConfig(object):
    """class that represents data associated with model:
    training data, testing data, checkpoints of trained model,
    outputs during training, etc."""
    root_results_dir = attr.ib(type=str)
    data_dir = attr.ib(type=str)
    module = attr.ib(type=str, default='mnist')
    train_size = attr.ib(converter=attr.converters.optional(float), default=None)
    val_size = attr.ib(converter=attr.converters.optional(float), default=None)
    # below gets added by main script to config file during training so it can be used
    # when measuring accuracy on test set
    results_dir_made_by_main = attr.ib(type=attr.converters.optional(str), default=None)


@attr.s
class TrainConfig(object):
    """class that represents configuration for training a RAM model"""
    batch_size = attr.ib(converter=int, default=10)
    learning_rate = attr.ib(converter=float, default=1e-3)
    epochs = attr.ib(converter=int, default=200)
    optimizer = attr.ib(type=str, default='momentum')
    beta1 = attr.ib(converter=float, default=0.9)
    beta2 = attr.ib(converter=float, default=0.999)
    epsilon = attr.ib(converter=float, default=1e-08)
    momentum = attr.ib(converter=float, default=0.9)
    replicates = attr.ib(converter=int, default=5)
    checkpoint_prefix = attr.ib(type=str, default='ckpt')
    restore = attr.ib(converter=strtobool, default='False')
    shuffle_each_epoch = attr.ib(converter=strtobool, default='True')
    save_examples_every = attr.ib(converter=int, default=25)
    num_examples_to_save = attr.ib(converter=int, default=9)
    save_loss = attr.ib(converter=strtobool, default='False')
    save_train_inds = attr.ib(converter=strtobool, default='False')
    # user does not specify current replicate, gets changed by main()
    current_replicate = attr.ib(type=int, default=None)


@attr.s
class TestConfig(object):
    """class that represents configuration for testing a RAM model"""
    save_examples = attr.ib(converter=strtobool, default='True')
    num_examples_to_save = attr.ib(converter=attr.converters.optional(int), default=None)


@attr.s
class MiscConfig(object):
    """class that represents configuration parameters that do not fit in sections above"""
    save_log = attr.ib(converter=strtobool, default='True')
    random_seed = attr.ib(converter=attr.converters.optional(int), default=None)


@attr.s
class Config(object):
    """class that represents configuration loaded from config.ini file

    Attributes
    ----------
    model : ModelConfig
        instance of ModelConfig class, represents configuration of model
    train : TrainConfig
        instance of TrainConfig class, represents configuration for training model
    data : DataConfig
        instance of DataConfig class, represent configuration for data associated
        with model (training, testing, outputs)
    test : TestConfig
        instance of TestConfig class, represent configuration for testing model
    """
    data = attr.ib(type=DataConfig)
    model = attr.ib(type=ModelConfig, default=ModelConfig())
    train = attr.ib(type=TrainConfig, default=TrainConfig())
    test = attr.ib(type=TestConfig, default=TestConfig())
    misc = attr.ib(type=MiscConfig, default=MiscConfig())


def parse_config(config_file):
    """read config.ini file with config parser,
    returns options from file loaded into an
    instance of Config class.

    Parameters
    ----------
    config_file : str
        Path to a config.ini file. If None, the
        default configuration is returned.
        Default is None.

    Returns
    -------
    config : instance of Config class
        with attributes that represent configuration for model, training, and data
    """
    if os.path.isfile(config_file):
        config = ConfigParser()
        config.read(config_file)
        sections = set(config.sections())
    else:
        raise FileNotFoundError(f'did not find config file: {config_file}')

    if sections != VALID_SECTIONS:
        if sections < VALID_SECTIONS:
            missing_sections = VALID_SECTIONS - sections
            raise ValueError(f'config missing section(s): {missing_sections}')
        elif not VALID_SECTIONS >= sections:
            extra_sections = sections - VALID_SECTIONS
            raise ValueError(f'sections in config not valid: {extra_sections}')

    for section in sections:
        options = set(config.options(section))
        valid_options = set(VALID_OPTIONS[section])
        # any options not defined go to default, so don't check if options < valid_options
        # but do check if valid_options is a superset of options,
        # i.e. test whether every element in options is in valid_options
        if not valid_options >= options:
            extra_options = options - valid_options
            raise ValueError(f'options in in config section {section} not valid: {extra_options}')
        if section == 'data':
            if 'root_results_dir' not in options:
                raise ValueError("config must specify 'root_results_dir' option")
            if 'data_dir' not in options:
                raise ValueError("config must specify 'data_dir' option")

    model_config = ModelConfig(**config['model'])
    train_config = TrainConfig(**config['train'])
    data_config = DataConfig(**config['data'])
    test_config = TestConfig(**config['test'])
    misc_config = MiscConfig(**config['misc'])
    config = Config(model=model_config, train=train_config, data=data_config, test=test_config, misc=misc_config)
    return config
